File: 2019/2019W28/preppindataw28.R
library(readxl)
library(dplyr)
library(tidyr)
library(stringr)
library(lubridate)

xlsx <- 'E:/PD Week 28 Input.xlsx'

data_head <- read_excel(xlsx, col_names = FALSE, n_max = 2) %>%
  mutate('col_level' = row_number()) %>%
  gather(., key='var', value='cols', -'col_level') %>%
  group_by(`col_level`) %>% 
  fill(`cols`) %>%
  spread(.,`col_level`, `cols`) %>%
  mutate('col_name' = ifelse(is.na(`2`),`1`,paste0(`1`,'-',`2`))) %>%
  arrange(as.integer(str_extract(`var`,'\\d+')))

final <- read_excel(xlsx, skip = 2 ,col_names = FALSE) %>%
  rename_all(funs(data_head[['col_name']][as.integer(str_extract(.,'\\d+'))])) %>%
  group_by(`Employee`) %>%
  arrange(`Observation Interval`) %>%
  mutate('Accum Length (mins)' = cumsum(`Observation Length (mins)`),
         'Observation Start Time' = dmy_hms(paste(str_extract(excel_sheets(xlsx)[1],'(\\d+\\.*)+'),
                                                  format(`Observation Start Time`,'%H:%M:%S'))) 
         %m+% minutes(x = `Accum Length (mins)`-`Observation Length (mins)`)) %>%
  ungroup() %>%
  gather(., key='Type Name', value='Type Value', contains('-'), na.rm = TRUE) %>%
  select(-c('Accum Length (mins)','Type Value')) %>%
  separate(`Type Name`, c('Real Col Name', 'Real Col Value'), sep = '-') %>%
  spread(.,`Real Col Name`, `Real Col Value`) %>%
  rename('Interaction' = 'Interaction With')

View(final)

File: 2019/2019W28/PreppindataW28.py
import pandas as pd
import re
from datetime import datetime as dt
from dateutil.relativedelta import relativedelta

xlsx = pd.ExcelFile(r"E:\PD Week 28 Input.xlsx")
sheet_name = xlsx.sheet_names[0]
data = pd.read_excel(xlsx,sheet_name,header=[0, 1]) 
data.columns = [col[0] if re.match('^Unnamed',col[1]) else col[0]+'-'+col[1] for col in data.columns.values]
data['Accum Length (mins)']=data.sort_values('Observation Interval').groupby(['Employee'])['Observation Length (mins)'].cumsum()
data['Observation Start Time']=data.apply(lambda x: dt.strptime(re.search('((\d+\.*)+)',sheet_name).group(0)+' '+str(x['Observation Start Time']),
    '%d.%m.%Y %H:%M:%S')+relativedelta(minutes=x['Accum Length (mins)']-x['Observation Length (mins)']),axis=1) 
data.drop('Accum Length (mins)',axis=1,inplace = True)
data = data.melt(id_vars=[c for c in data.columns if not re.match('(.*-.*)',c)],
                          value_name='Type Value', var_name='Type Name').dropna().drop('Type Value',axis=1)
temp = data['Type Name'].str.split('-',n = 1, expand = True)
data['Real Col Name'] = temp[0].apply(lambda x: re.sub('\sWith$','',x))
data['Real Col Value'] = temp[1]
data.drop('Type Name',axis=1,inplace = True)
final = data.pivot_table(index=[c for c in data.columns if not re.match('^Real',c)],
                                columns='Real Col Name', values='Real Col Value', aggfunc=max)
final.reset_index(inplace=True)
